Clarify documentation for `versioning-strategy` option (#20732)
github/docs: content/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file.md

            <td><code>^1.0.0</code></td>
        </tr>
      <tr>
            <td rowspan="3" align="center"><code>^1.0.0</code></td>
            <td rowspan="3" align="center"><code>1.0.0</code></td>
            <td rowspan="3" align="center"><code>2.0.0</code></td>
            <td align="center"><code>'widen'</code></td>
            <td><code>&gt;=1.0.0 &lt;3.0.0</code></td>
        </tr>
        <tr>
            <td align="center"><code>'increase'</code></td>
            <td><code>^2.0.0</code></td>
        </tr>
        <tr>
            <td align="center"><code>'increase-if-necessary'</code></td>
            <td><code>^2.0.0</code></td>
        </tr>
    </tbody>
</table>

Use the `versioning-strategy` option to change this behavior for supported package managers.

{% data reusables.dependabot.option-affects-security-updates %}

Available update strategies

| Option | Supported by | Action |
|--------|--------------|--------|
| `lockfile-only` | `bundler`, `cargo`, `composer`, `mix`, `npm`, `pip` | Only create pull requests to update lockfiles. Ignore any new versions that would require package manifest changes. |

> | `auto` | `bundler`, `cargo`, `composer`, `mix`, `npm`, `pip` | Follow the default strategy described above.|

Author: I was a little confused by the old docs... they imply all these ecosystems support `auto` which implies these ecosystems also support `widen` / `increase`... but not all these ecosystems are listed under `widen` / `increase`... is that an omission or am I misunderstanding how `auto` works?

Author: The code was difficult to trace, so I finally restored to testing this by setting up a test repo and trying different options. Thankfully the `dependabot.yml` configuration validator CI check would not only flag wrong values, but also list the allowed values, so it wasn't quite as painful as it could have been.

> | Ecosystem | Supported _versioning-strategies_ | Default strategy | Additional notes |

Author: Previously the docs implied that the default strategy is `auto` for _all_ ecosystems... However, regardless of whether true today, that may not be true in the future, for example https://github.com/dependabot/dependabot-core/issues/4979 so I added the `Default` column

> | `bundler` | `auto`, `increase`, `increase-if-necessary`, `lockfile-only` | `auto` |

Reviewer: `increase` and `increase-if-necessary` are misnamed. I can't find any record of them in our codebase.

I think these were meant to be `bump_versions` and `bump_versions_if_necessary`, which we support in [our dry-runner](https://github.com/dependabot/dependabot-core/blob/e11608fd17da3275cefca5e22eb01d01eb62fe09/bin/dry-run.rb#L196-L197) and were added to the bundler ecosystem in [Dependabot v0.85.2](https://github.com/dependabot/dependabot-core/blob/e11608fd17da3275cefca5e22eb01d01eb62fe09/CHANGELOG_ARCHIVE_2017_TO_2018.md?plain=1#L96-L99)

Author: 🤦 

I'm surprised that the incorrect config values weren't blowing up...probably they were silently getting swallowed as no-ops. I'll update this.

Reviewer 2: Actually turns out the `increase-if-necessary` and `increase` terminology is correct. The `bump_versions` and `bump_versions_if_necessary` are just internal naming, but the proper values for the configuration are the ones documented 👍.

| `gitsubmodule` | TODO | TODO |
| `go` | TODO | TODO |
| `gradle` | TODO | TODO |
| `maven` | TODO | TODO |
| `mix` | `auto`, `lockfile-only` | `auto` |
| `npm` | `auto`, `lockfile-only`, `widen` | `auto` |
| `nuget` | TODO | TODO |

> | `pip` | `auto`, `lockfile-only` | `auto` |

Reviewer 3: ```suggestion
| `pip` | `auto`, `increase`, `increase-if-necessary`, `lockfile-only`, `widen` | `auto` |
```
`increase` is definitely supported here, and with #6632 merged I think `widen` can be added as well?
`increase-if-necessary` I've never used myself, but from the error message I got when I tried to use `widen` I think it's supported as well:
`The property '#/updates/0/versioning-strategy' value "widen" did not match one of the following values: lockfile-only, auto, increase, increase-if-necessary`

Reviewer 2: Yes, all those 4 are currently supported for Python projects, and I plan to work on adding `widen` soon.

I think you meant to link to https://github.com/dependabot/dependabot-core/pull/6632, which fixed the update logic to better support `widen`, but we still need some internal work to expose the option in the configuration file.

Reviewer 3: I see, thanks for the correction. I updated the suggested change and the link in my comment.

Author: Yeah, I just tried on a test repo, and `widen` isn't yet supported, or at least the config file validator chokes on it. Thanks to @deivid-rodriguez 's hard work, some of the building blocks are in place, but there's a little more to do to get it across the finish line.

| `pub` | TODO | TODO |

> | `terraform` | TODO | TODO |

Author: These TODOs need to be filled out before merge. I work on the Dependabot team, so I can drive that.

Author: I both spelunked in the code and also created a test repository... what I've got are the currently supported values.

**Note:** The strategy code is open source, so if you'd like a particular ecosystem to support a new strategy, you are always welcome to submit a PR to https://github.com/dependabot/dependabot-core/.

```yaml
# Example config for customizing the manifest version strategy

version: 2
updates:
  - package-ecosystem: "composer"
    directory: "/"
    schedule:

>       interval: "weekly"

Author: Changing the interval was intentional... see https://github.com/github/docs/pull/20734/

### `versioning-strategy`

When {% data variables.product.prodname_dependabot %} edits a manifest file to update a version, there are several different potential _versioning-strategies_:

> | Option | Supported by | Action |

Author: Haha, great catch! I meant to delete it as part of the point of this PR is that users typically want to index by ecosystem => strategy, rather than the other way around. Fixed!

|--------|--------------|--------|
| `auto` | Try to differentiate between between apps and libraries. Use `increase` for apps and `widen` for libraries.|
| `increase`| Always increase the minimum version requirement to match the new version. If a range already exists, typically this only increases the lower bound. |
| `increase-if-necessary` | Leave the constraint if the original constraint allows the new version, otherwise, bump the constraint. |
| `lockfile-only` | Only create pull requests to update lockfiles. Ignore any new versions that would require package manifest changes. |

> | `widen`| Widen the allowed version requirements to include both the new and old versions, when possible. Typically this only increases the maximum allowed version requirement. |

Reviewer 4: `Typically` requires a comma when placed at the start of a sentence.
```suggestion
| `widen`| Widen the allowed version requirements to include both the new and old versions, when possible. Typically, this only increases the maximum allowed version requirement. |
```

| N/A | Some package managers do not yet support configuring the `versioning-strategy` parameter. |

> Example:

Reviewer 4: Following our content model:
```suggestion
The following table shows and example of how `versioning-strategy` can be used.
```

Author: 👍 but let me fix the typo of `"shows and example"` 😁

Reviewer 4: Argh, so sorry about that 🙈

> <table>

Reviewer 4: Can we please use markdown for this table?

Author: The problem I ran into was I couldn't find a clean way to do nested tables with markdown... see the original place I grabbed this table from here in this post: https://github.com/dependabot/dependabot-core/issues/4979#issue-1196967453

I poked around stackoverflow and most of the answer recommended either HTML or a custom markdown flavor/extension, which probably isn't supported here.

I do think it renders better as a nested table, so I'm not sure what to do here.

Can you handle formatting it according to docs guidance? I'm not sure if you'll still want to do left-aligned for a nested table or some variation, so just do whatever docs team prefers.

Author: I just realized I'd opened this from my personal fork of docs 🤦 which can be painful to deal with as a maintainer even though I've given maintainers edit access, so if it's too annoying feel free to leave a comment with your update (or just ```suggestion I suppose works quick!)

Reviewer 4: I am going to merge this great work, and will do the table conversion internally, so that we have the updates in our public-facing docs as soon as possible for our customers.

            <td><code>^1.0.0</code></td>
        </tr>
        <tr>
            <td align="center"><code>'increase'</code></td>
            <td><code>^1.2.0</code></td>
        </tr>
        <tr>
            <td align="center"><code>'increase-if-necessary'</code></td>
            <td><code>^1.0.0</code></td>
        </tr>
      <tr>
            <td rowspan="3" align="center"><code>^1.0.0</code></td>
            <td rowspan="3" align="center"><code>1.0.0</code></td>
            <td rowspan="3" align="center"><code>2.0.0</code></td>
            <td align="center"><code>'widen'</code></td>
            <td><code>&gt;=1.0.0 &lt;3.0.0</code></td>
        </tr>
        <tr>
            <td align="center"><code>'increase'</code></td>
            <td><code>^2.0.0</code></td>
        </tr>
        <tr>
            <td align="center"><code>'increase-if-necessary'</code></td>
            <td><code>^2.0.0</code></td>
        </tr>
    </tbody>
</table>

Use the `versioning-strategy` option to change this behavior for supported package managers.

{% data reusables.dependabot.option-affects-security-updates %}

Available update strategies:

> | Ecosystem | Supported _versioning-strategies_ | Default strategy | Additional notes |

Reviewer 4: ```suggestion
| Ecosystem | Supported versioning strategies | Default strategy | Additional notes |
```

|-----------|-----------------------------------|------------------|------------------|

> | `bundler` | `auto`, `increase`, `increase-if-necessary`, `lockfile-only` | `auto` |

Reviewer 4: We don't seem to really need the "Additional notes" column as it seems to be a repeat of the `Default strategy` column, unless I misunderstood its purpose. Shall we remove it?

| `cargo` | `auto`, `lockfile-only` | `auto` |
| `composer` | `auto`, `increase`, `increase-if-necessary`, `lockfile-only`, `widen` | `auto` |
| `docker` | N/A | N/A |
| `github-actions` | N/A | N/A |
| `gitsubmodule` | N/A | N/A |
| `gomod` | | N/A | N/A |
| `gradle` | | N/A | N/A |
| `maven` | | N/A | N/A |
| `mix` | `auto`, `lockfile-only` | `auto` |
| `npm` | `auto`, `increase`, `increase-if-necessary`, `lockfile-only`, `widen` | `auto` |
| `nuget` | | N/A | N/A |
| `pip` | `auto`, `increase`, `increase-if-necessary`, `lockfile-only` | `auto` |
| `pub` | `auto`, `increase`, `increase-if-necessary`, `widen` | `auto` |
| `terraform` | | N/A | N/A |

> **Note:** `N/A` indicates the package manager does not yet support configuring the `versioning-strategy` parameter. The strategy code is open source, so if you'd like a particular ecosystem to support a new strategy, you are always welcome to submit a PR to https://github.com/dependabot/dependabot-core/.

Reviewer 4: Typo + following the correct syntax for a note box.

```suggestion
{% note %}

**Note:** `N/A` indicates that the package manager does not yet support configuring the `versioning-strategy` parameter. The strategy code is open source, so if you'd like a particular ecosystem to support a new strategy, you are always welcome to submit a pull request in https://github.com/dependabot/dependabot-core/.

{% endnote %}
```

Reviewer 5: > Typo + following the correct syntax for a note box.

```yaml

> # Example config for customizing the manifest version strategy

Reviewer 4: ```suggestion
# Example configuration for customizing the manifest version strategy
```
